From 121a3493ffb2db869ade34bad6f33e90e7ce0cd4 Mon Sep 17 00:00:00 2001
From: Rajat Soni <quic_rajson@quicinc.com>
Date: Wed, 7 Jun 2023 11:49:23 +0530
Subject: [PATCH] ath12k: Add soc_dp_stats for incorrect reo desc first and last msdu

Currently we are receiving a ppdu that is not having first/last msdu bit set in
reo desc i.e hal_reo_dest_ring this causes this warning.

Due to this no functionality is impacted, so moving this warning to stats to get
how many times first/last msdu bit is not set.

Note: This is a workaround need to revert this change.

Signed-off-by: Rajat Soni <quic_rajson@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h    | 1 +
 drivers/net/wireless/ath/ath12k/debugfs.c | 4 ++++
 drivers/net/wireless/ath/ath12k/dp_rx.c   | 4 +++-
 3 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 61bf541..b160162 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -1138,6 +1138,7 @@ struct ath12k_soc_dp_stats {

	struct ath12k_soc_dp_tx_err_stats tx_err;
	u32 reo_cmd_update_rx_queue_error;
+	u32 first_and_last_msdu_bit_miss;
 };
 
 #ifdef CPTCFG_ATH12K_PPE_DS_SUPPORT
diff --git a/drivers/net/wireless/ath/ath12k/debugfs.c b/drivers/net/wireless/ath/ath12k/debugfs.c
index 7d6c33f..a88c9ce 100644
--- a/drivers/net/wireless/ath/ath12k/debugfs.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs.c
@@ -1755,6 +1755,10 @@ static ssize_t ath12k_debugfs_dump_soc_dp_stats(struct file *file,
 				soc_stats->rx_wbm_rel_source[i][1],
 				soc_stats->rx_wbm_rel_source[i][2]);
 
+	len += scnprintf(buf + len, size - len,
+			"\nFIRST/LAST MSDU BIT MISSING COUNT: %u\n",
+			soc_stats->first_and_last_msdu_bit_miss);
+
 	len += ath12k_debugfs_dump_soc_ring_bp_stats(ab, buf + len, size - len);
 
 	if (len > size)
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index 1625805..7b0b932 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -2543,7 +2543,9 @@ static void ath12k_dp_rx_h_undecap_raw(struct ath12k *ar, struct sk_buff *msdu,
 
 	if (!rxcb->is_first_msdu ||
 	    !(rxcb->is_first_msdu && rxcb->is_last_msdu)) {
-		WARN_ON_ONCE(1);
+		/* TODO: Change below stats increment back to WARN_ON_ONCE(1)
+		*/
+		ar->ab->soc_stats.first_and_last_msdu_bit_miss++;
 		return;
 	}
 
-- 
2.17.1

